File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.py
def solution(s):
    s = list(s)
    if s[-1] == "(":
        return False
    
    tmp = []
    while s:
        if s[-1] == "(" and tmp:
            tmp.pop()
            s.pop()
        elif s[-1] == ")":
            tmp.append(s.pop())
        else:
            return False
    
    if not tmp and not s:
        return True
    return False

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.*;

class Solution {
    boolean solution(String s) {
        boolean answer = true;
        int count = 0;
        
        Deque<Character> q = new ArrayDeque<>();
        for (int i=0;i<s.length();i++)
            q.push(s.charAt(i));

        char now = q.pop();
        if (now=='(') return false;
        else count++;
        
        while (!q.isEmpty()) {
            if (q.peek()=='(') {
                if (count > 0) {
                    count--;
                } else return false;
            } else count++;
            
            q.pop();
        }
        
        if (count != 0) return false;
        return answer;
    }
}